### ISSUE TITLE
including internal light sources

### ISSUE THREAD

**Issue by swathimurthy9** [InverseLight/ValoMC#21]
Hello, 

I have been using ValoMC for light simulation in biofilms. We recently published a paper using the tool: 
https://doi.org/10.1098/rsif.2021.0532

We would like to use the tool for another project to simulate auto-fluorescence in biofilms. For this we like to include internal isotropic light sources in certain domains in the geometry. Could you point out how this can be achieved? Or the possibility for building this together?

Bw
Swathi

**Comment by aaleino** [comment]
Hi! Would you like to have a point-like sources in the domain (all photons emitted from the same location) or a volumetric source (e.g. photons emitted randomly from volume elements)? In 2D or 3D? Br, Aleksi

**Comment by aaleino** [comment]
I have drafted such light sources in the git branch "volumetric_lights". But I haven't tested it. I think next step would be to build some scenario that has an analytical solution, and compare the results of the code to the solution. Otherwise, it is very risky to use the code.

**Comment by swathimurthy9** [comment]
Thank you for including the volumetric light source. I tried to run the code on a 3D geometry consisting of a fluorescent slab surrounded by a non-fuoroscent box, for 3 scenarios:

1. no fluorescence
2. 10% fluorescence efficiency
3. 60% fluorescence efficiency

I have attached the results, normalized to the fluence rate of the exciting light source, as a screen shot here. The results look a bit confusing. At 10% efficiency , the fluence in the surrounding box is close to 1, where as at 60% efficiency it is 0.4. Does the solution.element_fluence include both the excitation light (distribution) and the flourescence? Is it possible to plot only the flourescence? 





![screen_shot](https://user-images.githubusercontent.com/109286900/185367068-00172b93-784d-4d0b-a803-a4422a1c5ac2.png)

**Comment by swathimurthy9** [comment]
The fluorescent slab is also scattering.

**Comment by aaleino** [comment]
Hi, sorry for the late response. The field "volumetric_light_source_fraction" is not the efficiency of the fluorescence, but simply the fraction of all photons that will be sent from the volumetric light source.  So if you want the contribution of only the volumetric light, set "volumetric_light_source_fraction" to one. Then run another simulation with only the boundary lights on "volumetric_light_source_fraction" zero, and add the results together with a suitable normalization. E.g. volumetric light is 6 times stronger (6 W) than boundary lights, multiply volumetric results by 6. 

So this new addition was simply adding volumetric lights, and a not direct simulation of auto-fluorescence. 

If you want to simulate the physical process of each photon having some probability to trigger fluorescence and get converted to a fluorescence photon, that's a more difficult problem, and I have to think about that. Is that what you need?

**Comment by swathimurthy9** [comment]
Thank you for the reply. Yes, I would like to simulate auto-fluorescence, as consequence of excitation light (originating from the boundary and scattered in the medium) being converted at a certain efficiency. On the lines described here:
https://omlc.org/software/mc/
https://omlc.org/software/mc/mcfluor/index.html

Would that be possible to implement in ValoMC?

**Comment by aaleino** [comment]
At the moment I do not have the time to do literature search, but if you can describe the  mathematics and workings of such algorithm here briefly we can see what can be done (unfortunately, without any guarantee of the timetable).  How is the efficiency defined? Out of the absorbed photons, certain fraction turn to fluorescence photons?

**Comment by swathimurthy9** [comment]
Yes, efficiency is defined as a certain fraction of the excitation light converted by fluorescence. 
[MC_fluor_equations.pdf](https://github.com/InverseLight/ValoMC/files/9497558/MC_fluor_equations.pdf)
I have attached the general description of the algorithm and the equations here. Please let me know if you need any further details. Thank you very much.

**Comment by aaleino** [comment]
The document was a bit abstract and I got bogged down in details when reading. To put very simply how I imagine this kind of simulation would work: instead of absorption coefficient, we could define "fluorescence absorption coefficient". This is the probability per unit length that an fluorescence occurs as the photon propagates. If fluorescence occurs, the photon will be re-emitted to (uniformly) random direction. Is that correct?

**Comment by swathimurthy9** [comment]
Thank you for the quick reply. I will try to explain here.

Let's consider that the geometry has 2 domains and the fluorescence occurs in only domain 1 from a uniform distribution of background fluorophore. There are 2 parts for the simulation:

MonteCarlo simulation of excitation light, using the optical properties of excitation wavelength. This gives a fluence rate distribution in both the domains.
Followed by MonteCarlo simulation of auto-fluorescence in domain 1 only, using the optical properties for fluorescence wavelength. Considering the fluorophores as internal, isotropic, light sources generated due to the excitation light.
The auto-fluorescence generation in each mesh element in Domain 1 depends on:
a. The fluence rate of the excitation light at the given location (mesh element).
b. We can use a lumped factor which implies amount of the excitation light converted to fluorescent light (fluorescence yield). Example: excitation fluence rate at a given location is 1 W/cm^2 per W of incident light and the lumped factor is 0.1. Then the intensity of the fluorescent source at that location is 0.1 W/cm^2 per W of incident light.

All the fluorescent light generated in domain 1 undergoes scattering and absorption depending on the optical properties defined for that wavelength (MC simulation). This finally gives a distribution of fluorescent light intensity in the geometry.

Please let me know if something is unclear.

Cheers
Swathi

**Comment by aaleino** [comment]
Just a status update: I haven't forgotten about this. I am just too busy at the moment to work on this!

**Comment by swathimurthy9** [comment]
Hello Aleksi Leino,

I was thinking of how to implement auto-fluorescence from the existing code you have already made for Volumetric light sources, Instead of assuming all the internal elements are generating photons uniformly (of the same intensity), we use the element_fluence from the solution for the excitation source and multiply this with 'fluorescence absorption co-efficient', which will give the intensity of the volumetric source as W/m^2/W of incident light (which was the initial boundary source). Is this difficult to implement?

Cheers
Swathi

**Comment by aaleino** [comment]
Hello, What you described in previous messages is possible to implement. You could already do this with the existing code by forming the second solution from a sum of several simulations in which elements were individually lighted up (assuming the code for the volumetric light works - it should be tested). 

I can start working on wrapping up a package for doing this in two simulations, but as I can work on this only very inconsistently and seldomly, it will take a while. 

So I assume the fluorescent light does not cause any more fluorescence? In other words, if we used the fluorescence wavelength in the "external light source", no fluorescence would occur? Otherwise, the system will not work.

**Comment by swathimurthy9** [comment]
Thanks for the reply. It's great to hear that it is possible to implement this with the existing code. Yes, we can assume that the fluorescent light does not cause any more fluorescence. 

I understand that you are busy and do not get much time to work on this. Thank you for your time.
I look forward to the auto-fluorescence package.

Cheers
Swathi

**Comment by swathimurthy9** [comment]
<img width="333" alt="grooves" src="https://github.com/user-attachments/assets/f41dcc18-a827-45b5-b92e-3a8d4287aa7a">

Attachment with axes.
